R/input-utils.R

```diff
@@ -90,31 +90,32 @@ generateOptions <- function(inputId, selected, inline, type = 'checkbox',
   div(class = "shiny-options-group", options)
 }
 
+# If a list/vector is unnamed, give it blank names
+ensureNamed <- function(x) {
+  if (is.null(names(x))) names(x) <- character(length(x))
+  x
+}
 
-# Takes a vector or list, and adds names (same as the value) to any entries
-# without names. Coerces all leaf nodes to `character`.
-choicesWithNames <- function(choices) {
-  # Take a vector or list, and convert to list. Also, if any children are
-  # vectors with length > 1, convert those to list. If the list is unnamed,
-  # convert it to a named list with blank names.
-  listify <- function(obj) {
-    # If a list/vector is unnamed, give it blank names
-    makeNamed <- function(x) {
-      if (is.null(names(x))) names(x) <- character(length(x))
+# Take a vector or list, and convert to list. Also, if any children are
+# vectors with length > 1, convert those to list. If the list is unnamed,
+# convert it to a named list with blank names.
+listify <- function(x) {
+  if (is.list(x) || is.null(x)) {
+    ensureNamed(lapply(x, listify))
+  } else if (is.character(x)) {
+    if (length(x) == 1 && is.null(names(x))) {
       x
+    } else {
+      as.list(ensureNamed(x))
     }
-
-    res <- lapply(obj, function(val) {
-      if (is.list(val))
-        listify(val)
-      else if (length(val) == 1 && is.null(names(val)))
-        as.character(val)
-      else
-        makeNamed(as.list(val))
-    })
-
-    makeNamed(res)
+  } else {
+    listify(stats::setNames(as.character(x), names(x)))
```

Comment:
It would help to have a comment saying that this is the code path where factors will end up (otherwise it's non-obvious what types take this path and we may forget in the future).